File: src/main/java/top/mcfpp/lang/Range.java
package top.mcfpp.lang;
public class Range<T extends MCNumber<?>> {

    /**
     * 闭区间左端
     */
    public T start;

    /**
     * 闭区间右端
     */
    public T end;


    /**
     * 构造一个区间
     * @param start 区间左端。若为null则没有左端。
     * @param end 区间右端。若为null则没有右端。
     */
    public Range(T start, T end){
        this.start = start;
        this.end = end;
    }



    @Override
    public String toString() {
        return  (start != null ? start : "") +  ".." + (end != null ? end : "");
    }
}

File: src/main/kotlin/top/mcfpp/lang/Range.kt
package top.mcfpp.lang

interface Range<T>{
    fun inRange(value: T): Boolean
    fun clone(): Range<T>
}

abstract class OneValueRange<T>(var value : T) : Range<T>{
    override fun inRange(value: T): Boolean {
        return value == this.value
    }
}

abstract class TwoValueRange<T>(var start: T?, var end: T?) : Range<T>{

    override fun inRange(value: T): Boolean {
        if(start == null && end == null) throw NullPointerException("Range should have at least one side")
        val re = if(start != null) leftValueCompare(value) else true
        return re && rightValueCompare(value)
    }

    abstract fun leftValueCompare(value: T) : Boolean

    abstract fun rightValueCompare(value: T) : Boolean

}

class OneIntConcreteRange(value: Int): OneValueRange<Int>(value){
    override fun clone(): Range<Int> {
        return OneIntConcreteRange(value)
    }
}

class TwoIntConcreteRange(start: Int?, end: Int?) : TwoValueRange<Int>(start, end) {
    override fun leftValueCompare(value: Int): Boolean {
        return value >= start!!
    }

    override fun rightValueCompare(value: Int): Boolean {
        return value <= end!!
    }

    override fun clone(): Range<Int> {
        return TwoIntConcreteRange(start, end)
    }
}

class OneFloatConcreteRange(value: Float) : OneValueRange<Float>(value){
    override fun clone(): Range<Float> {
        return OneFloatConcreteRange(value)
    }
}

class TwoFloatConcreteRange(start: Float?, end: Float?) : TwoValueRange<Float>(start, end) {
    override fun leftValueCompare(value: Float): Boolean {
        return value >= start!!
    }

    override fun rightValueCompare(value: Float): Boolean {
        return value <= end!!
    }

    override fun clone(): Range<Float> {
        return TwoFloatConcreteRange(start, end)
    }
}